openfeature/multi/isolation.go

```diff
@@ -0,0 +1,334 @@
+// return of this function is a merged version of the evaluation context
+func (h *hookIsolator) beforeHooks(ctx context.Context) (of.EvaluationContext, error) {
+	contexts := []of.EvaluationContext{h.capturedContext.EvaluationContext()}
+	for _, hook := range h.hooks {
+		resultEvalCtx, err := hook.Before(ctx, h.capturedContext, h.capturedHints)
+		if resultEvalCtx != nil {
+			contexts = append(contexts, *resultEvalCtx)
+		}
+		if err != nil {
+			return mergeContexts(contexts...), err
+		}
+	}
+
+	return mergeContexts(contexts...), nil
+}
+
+// afterHooks executes all after [of.Hook] instances together.
+func (h *hookIsolator) afterHooks(ctx context.Context, evalDetails of.InterfaceEvaluationDetails) error {
+	for _, hook := range h.hooks {
+		if err := hook.After(ctx, h.capturedContext, evalDetails, h.capturedHints); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+// errorHooks executes all error [of.Hook] instances together.
+func (h *hookIsolator) errorHooks(ctx context.Context, err error) {
+	for _, hook := range h.hooks {
+		hook.Error(ctx, h.capturedContext, err, h.capturedHints)
+	}
+}
+
+// finallyHooks execute all finally [of.Hook] instances together.
+func (h *hookIsolator) finallyHooks(ctx context.Context, details of.InterfaceEvaluationDetails) {
+	for _, hook := range h.hooks {
+		hook.Finally(ctx, h.capturedContext, details, h.capturedHints)
+	}
+}
+
+// updateEvalContext returns a new [of.HookContext] with an updated [of.EvaluationContext] value. [of.HookContext] is
+// immutable, and this returns a new [of.HookContext] with all other values copied
+func (h *hookIsolator) updateEvalContext(evalCtx of.EvaluationContext) {
+	hookCtx := of.NewHookContext(
+		h.capturedContext.FlagKey(),
+		h.capturedContext.FlagType(),
+		h.capturedContext.DefaultValue(),
+		h.capturedContext.ClientMetadata(),
+		h.capturedContext.ProviderMetadata(),
+		evalCtx,
+	)
+	h.mu.Lock()
+	defer h.mu.Unlock()
+	h.capturedContext = hookCtx
+}
+
+// deepenContext converts a [of.FlattenedContext] to a [of.EvaluationContext].
+func deepenContext(flatCtx of.FlattenedContext) of.EvaluationContext {
+	noTargetingKey := make(map[string]any)
+	for k, v := range flatCtx {
+		if k != of.TargetingKey {
+			noTargetingKey[k] = v
+		}
+	}
+	var targetingKey string
+	if tk, ok := flatCtx[of.TargetingKey]; ok {
+		targetingKey, _ = tk.(string)
+	}
+	return of.NewEvaluationContext(targetingKey, noTargetingKey)
+}
+
+// flattenContext converts a [of.EvaluationContext] to a [of.FlattenedContext]
+func flattenContext(evalCtx of.EvaluationContext) of.FlattenedContext {
+	flatCtx := evalCtx.Attributes()
+	flatCtx[of.TargetingKey] = evalCtx.TargetingKey()
+	return flatCtx
+}
+
+// mergeContexts merges attributes from the given EvaluationContexts with the nth [of.EvaluationContext] taking precedence
+// in case of any conflicts with the (n+1)th [of.EvaluationContext].
+func mergeContexts(evaluationContexts ...of.EvaluationContext) of.EvaluationContext {
```

Comment:
There is a mergeContexts func in openfeature/client.go that you can use instead:
https://github.com/open-feature/go-sdk/blob/40fe27eca801dc7ffb4865e403bc2f407fddc339/openfeature/client.go#L836